tango/step_caches/local_step_cache.py

```diff
@@ -68,44 +90,54 @@ def _get_from_cache(self, key: str) -> Optional[Any]:
         except KeyError:
             return None
 
+    def _metadata_path(self, step_or_unique_id: Union[Step, StepInfo, str]) -> Path:
+        return self.step_dir(step_or_unique_id) / self.METADATA_FILE_NAME
+
     def __contains__(self, step: object) -> bool:
-        if isinstance(step, Step) and step.cache_results:
+        if (isinstance(step, Step) and step.cache_results) or (
+            isinstance(step, StepInfo) and step.cacheable
+        ):
             key = step.unique_id
             if key in self.strong_cache:
                 return True
             if key in self.weak_cache:
                 return True
-            metadata_file = self.step_dir(step) / "cache-metadata.json"
-            return metadata_file.exists()
+            return self._metadata_path(
+                cast(Union[Step, StepInfo], step)  # cast is for mypy :/
+            ).exists()
         else:
             return False
 
-    def __getitem__(self, step: Step) -> Any:
+    def __getitem__(self, step: Union[Step, StepInfo]) -> Any:
         key = step.unique_id
         result = self._get_from_cache(key)
         if result is None:
             if step not in self:
                 raise KeyError(step)
-            result = step.format.read(self.step_dir(step))
+            metadata = CacheMetadata.from_params(Params.from_file(self._metadata_path(step)))
+            result = metadata.format.read(self.step_dir(step))
             self._add_to_cache(key, result)
         return result
 
     def __setitem__(self, step: Step, value: Any) -> None:
         if not step.cache_results:
-            logger.warning("Tried to cache step %s despite being marked as uncacheable.", step.name)
+            warnings.warn(
+                f"Tried to cache step '{step.name}' despite being marked as uncacheable",
+                UserWarning,
+            )
```

Comment:
The `logger` construction has the advantage that it won't do the string substitution unless the string is actually needed (i.e., warnings are not silenced). A `f""` string is always created first, before we find out whether the string is needed.

Reply:
I'll bet any potential perf benefits of `logger.warning` vs `warnings.warn` are completely negligible in this case, and in my opinion usage warnings from a library should always be emitted via the `warnings` module so that warnings will propagate even if logging isn't configured.